### PR TITLE
Adding debug capabilities for client

### PR DESCRIPTION
**Pull request by vishalnarkhede**
# Submit a pull request

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

## Description of the pull request

Consumer can get logs by configuring logger following way:

```
   const chatClient = new StreamChat('api_key', {}, {
      logger: (level, msg, extra={}) => {
          if (extra.tags.indexOf('api_request') > -1) {
            console.log(msg);
          }
      },
    });

```

### CODE REVIEW

**Comment by tbarbugli** [comment]
feels a little bit too much like Java meets JS. I would go with something more minimalistic. ie a logger is just a function like this `(log_level, msg, data) => {}`

**Comment by vishalnarkhede** [comment]
@tbarbugli changes have been done as discussed :)

File: src/client.js [hunk 1]
```diff
@@ -357,9 +413,38 @@ export class StreamChat {
...
+			url,
+			payload: data,
+			config,
+		});
+	}
+
+	_logApiResponse(type, url, response) {
+		this.logger(
+			'info',
+			'client:' +
```
Comment (tbarbugli) [inline src/client.js]: @vishalnarkhede any reason not to use string interpolation? this looks like 💩

File: src/client.js [hunk 2]
```diff
@@ -536,6 +641,14 @@ export class StreamChat {
 	}
 
 	recoverState = async () => {
+		this.logger(
+			'info',
+			'client:recoverState() - Begining of recoverState with connectionID' +
```
Comment (tbarbugli) [inline src/client.js]: begining -> beginning or even better start

File: src/client.js [hunk 3]
```diff
@@ -548,11 +661,24 @@ export class StreamChat {
 			lastMessageIDs[c.cid] = lastMessageId;
 		}
 		if (cids.length) {
+			this.logger(
+				'info',
+				'client:recoverState() - Starting the querying of ' +
```
Comment (tbarbugli) [inline src/client.js]: start or begin, pick one :)

File: src/channel.js [hunk 4]
```diff
@@ -409,6 +409,15 @@ export class Channel {
...
 		this.data = state.channel;
+
+		this._client.logger(
+			'info',
+			'channel:watch() - started watching channel ' + this.cid,
+			{
+				tags: ['channel'],
+				channel: this,
+			},
+		);
```
Comment (tbarbugli) [inline src/channel.js]: we should attach events to logging as well ie.

```js
client.on((e) => {client.logger(...)})
```

File: src/client.js (hunk 1 above)
Comment (vishalnarkhede) [reply src/client.js]: Done!

File: src/channel.js (hunk 4 above)
Comment (vishalnarkhede) [reply src/channel.js]: I have added a log in`.on(...)` now. Although it only attaches the listeners. When event is received,  `_handleChannelEvent ` and `_handleClientEvent ` are the functions that get called. I have already added logs to those functions :)

**Pull request merged**